// angular-material-fileupload/esm5/lib/file-upload-input-for/file-upload-input-for.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/file-upload-input-for/file-upload-input-for.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, EventEmitter, HostListener, Input, Output, } from "@angular/core";
/**
 * A material design file upload queue component.
 */
var FileUploadInputForDirective = /** @class */ (function () {
    function FileUploadInputForDirective(element) {
        this.element = element;
        this._queue = null;
        this.onFileSelected = new EventEmitter();
        this._element = this.element.nativeElement;
    }
    Object.defineProperty(FileUploadInputForDirective.prototype, "fileUploadQueue", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value) {
                this._queue = value;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FileUploadInputForDirective.prototype.onChange = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var files = this.element.nativeElement.files;
        this.onFileSelected.emit(files);
        for (var i = 0; i < files.length; i++) {
            this._queue.add(files[i]);
        }
        this.element.nativeElement.value = "";
    };
    /**
     * @param {?} event
     * @return {?}
     */
    FileUploadInputForDirective.prototype.onDrop = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var files = event.dataTransfer.files;
        this.onFileSelected.emit(files);
        for (var i = 0; i < files.length; i++) {
            this._queue.add(files[i]);
        }
        event.preventDefault();
        event.stopPropagation();
        this.element.nativeElement.value = "";
    };
    /**
     * @param {?} event
     * @return {?}
     */
    FileUploadInputForDirective.prototype.onDropOver = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
    };
    FileUploadInputForDirective.decorators = [
        { type: Directive, args: [{
                    selector: "input[fileUploadInputFor], div[fileUploadInputFor]",
                },] }
    ];
    /** @nocollapse */
    FileUploadInputForDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    FileUploadInputForDirective.propDecorators = {
        onFileSelected: [{ type: Output }],
        fileUploadQueue: [{ type: Input, args: ["fileUploadInputFor",] }],
        onChange: [{ type: HostListener, args: ["change",] }],
        onDrop: [{ type: HostListener, args: ["drop", ["$event"],] }],
        onDropOver: [{ type: HostListener, args: ["dragover", ["$event"],] }]
    };
    return FileUploadInputForDirective;
}());
export { FileUploadInputForDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FileUploadInputForDirective.prototype._queue;
    /**
     * @type {?}
     * @private
     */
    FileUploadInputForDirective.prototype._element;
    /** @type {?} */
    FileUploadInputForDirective.prototype.onFileSelected;
    /**
     * @type {?}
     * @private
     */
    FileUploadInputForDirective.prototype.element;
}
