// angular-material-fileupload/bundles/angular-material-fileupload.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/material/icon'), require('@angular/material/progress-bar'), require('@angular/material/card'), require('@angular/material/button'), require('@angular/common'), require('@angular/common/http'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('angular-material-fileupload', ['exports', '@angular/core', '@angular/material/icon', '@angular/material/progress-bar', '@angular/material/card', '@angular/material/button', '@angular/common', '@angular/common/http', 'rxjs', 'rxjs/operators'], factory) :
    (global = global || self, factory(global['angular-material-fileupload'] = {}, global.ng.core, global.ng.material.icon, global.ng.material['progress-bar'], global.ng.material.card, global.ng.material.button, global.ng.common, global.ng.common.http, global.rxjs, global.rxjs.operators));
}(this, (function (exports, core, icon, progressBar, card, button, common, http, rxjs, operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/bytes/bytes.pipe.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var BytesPipe = /** @class */ (function () {
        function BytesPipe() {
        }
        /**
         * @param {?} bytes
         * @return {?}
         */
        BytesPipe.prototype.transform = /**
         * @param {?} bytes
         * @return {?}
         */
        function (bytes) {
            if (isNaN(parseFloat('' + bytes)) || !isFinite(bytes))
                return '-';
            if (bytes <= 0)
                return '0';
            /** @type {?} */
            var units = ['bytes', 'KB', 'MB', 'GB', 'TB', 'PB'];
            /** @type {?} */
            var number = Math.floor(Math.log(bytes) / Math.log(1024));
            return (bytes / Math.pow(1024, Math.floor(number))).toFixed(1) + ' ' + units[number];
        };
        BytesPipe.decorators = [
            { type: core.Pipe, args: [{ name: 'bytes' },] }
        ];
        return BytesPipe;
    }());

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/mat-file-upload-queue/mat-file-upload-queue.service.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MatFileUploadQueueService = /** @class */ (function () {
        function MatFileUploadQueueService() {
            this.inputValueSubject = new rxjs.BehaviorSubject(null);
            this.inputValue$ = this.inputValueSubject.asObservable();
        }
        /**
         * @param {?} input
         * @return {?}
         */
        MatFileUploadQueueService.prototype.initialize = /**
         * @param {?} input
         * @return {?}
         */
        function (input) {
            this.inputValueSubject.next(input);
        };
        /**
         * @return {?}
         */
        MatFileUploadQueueService.prototype.getInputValue = /**
         * @return {?}
         */
        function () {
            return this.inputValueSubject.getValue();
        };
        MatFileUploadQueueService.decorators = [
            { type: core.Injectable }
        ];
        return MatFileUploadQueueService;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        MatFileUploadQueueService.prototype.inputValueSubject;
        /** @type {?} */
        MatFileUploadQueueService.prototype.inputValue$;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/mat-file-upload/mat-file-upload.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MatFileUploadComponent = /** @class */ (function () {
        function MatFileUploadComponent(HttpClient, matFileUploadQueueService) {
            this.HttpClient = HttpClient;
            this.matFileUploadQueueService = matFileUploadQueueService;
            this.uploadProgressSubject = new rxjs.ReplaySubject();
            this.uploadProgress$ = this.uploadProgressSubject.asObservable();
            this.uploadInProgressSubject = new rxjs.BehaviorSubject(false);
            this.uploadInProgress$ = this.uploadInProgressSubject.asObservable();
            this.subs = new rxjs.Subscription();
            this.fileUploadAriaLabel = "File Upload";
            this.cancelAriaLabel = "Cancel File Upload";
            /**
             * Output
             */
            this.removeEvent = new core.EventEmitter();
            this.onUpload = new core.EventEmitter();
            /** @type {?} */
            var queueInput = this.matFileUploadQueueService.getInputValue();
            if (queueInput) {
                this.httpUrl = this.httpUrl || queueInput.httpUrl;
                this.httpRequestHeaders =
                    this.httpRequestHeaders || queueInput.httpRequestHeaders;
                this.httpRequestParams =
                    this.httpRequestParams || queueInput.httpRequestParams;
                this.fileAlias = this.fileAlias || queueInput.fileAlias;
            }
        }
        Object.defineProperty(MatFileUploadComponent.prototype, "file", {
            get: /**
             * @return {?}
             */
            function () {
                return this._file;
            },
            set: /**
             * @param {?} file
             * @return {?}
             */
            function (file) {
                this._file = file;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(MatFileUploadComponent.prototype, "id", {
            get: /**
             * @return {?}
             */
            function () {
                return this._id;
            },
            set: /**
             * @param {?} id
             * @return {?}
             */
            function (id) {
                this._id = id;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        MatFileUploadComponent.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
            this.uploadProgressSubject.next({
                progressPercentage: 0,
                loaded: 0,
                total: this._file.size,
            });
        };
        /**
         * @return {?}
         */
        MatFileUploadComponent.prototype.upload = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.uploadInProgressSubject.next(true);
            // How to set the alias?
            /** @type {?} */
            var formData = new FormData();
            formData.set(this.fileAlias, this._file, this._file.name);
            this.subs.add(this.HttpClient.post(this.httpUrl, formData, {
                headers: this.httpRequestHeaders,
                observe: "events",
                params: this.httpRequestParams,
                reportProgress: true,
                responseType: "json",
            }).subscribe((/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                if (event.type === http.HttpEventType.UploadProgress) {
                    _this.uploadProgressSubject.next({
                        progressPercentage: Math.floor((event.loaded * 100) / event.total),
                        loaded: event.loaded,
                        total: event.total,
                    });
                }
                _this.onUpload.emit({ file: _this._file, event: event });
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                if (_this.fileUploadSubscription) {
                    _this.fileUploadSubscription.unsubscribe();
                }
                _this.uploadInProgressSubject.next(false);
                _this.onUpload.emit({ file: _this._file, event: event });
            }), (/**
             * @return {?}
             */
            function () { return _this.uploadInProgressSubject.next(false); })));
        };
        /**
         * @return {?}
         */
        MatFileUploadComponent.prototype.remove = /**
         * @return {?}
         */
        function () {
            this.subs.unsubscribe();
            this.removeEvent.emit(this);
        };
        /**
         * @return {?}
         */
        MatFileUploadComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
        function () {
            this.subs.unsubscribe();
        };
        MatFileUploadComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "mat-file-upload",
                        template: "<ng-container *ngIf=\"uploadProgress$ | async as uploadProgress\">\n  <mat-card>\n    <span class=\"file-summary\">{{ file.name }}({{ file.size | bytes }})</span>\n    <div class=\"upload-progress\">\n      <mat-progress-bar\n        [value]=\"uploadProgress.progressPercentage\"\n      ></mat-progress-bar>\n\n      <button\n        mat-icon-button\n        [attr.aria-label]=\"fileUploadAriaLabel\"\n        (click)=\"upload()\"\n        [disabled]=\"uploadInProgress$ | async\"\n      >\n        <mat-icon>file_upload</mat-icon>\n      </button>\n\n      <button\n        mat-icon-button\n        [attr.aria-label]=\"cancelAriaLabel\"\n        (click)=\"remove()\"\n      >\n        <mat-icon>cancel</mat-icon>\n      </button>\n    </div>\n    <span class=\"file-summary\">{{ uploadProgress.progressPercentage }}%</span>\n    <span>\n      {{ uploadProgress.loaded | bytes }} of\n      {{ uploadProgress.total | bytes }}</span\n    >\n  </mat-card>\n</ng-container>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        styles: [".file-summary{font-size:.85rem}.upload-progress{display:flex;align-content:center;align-items:center;height:10px}.upload-progress ::ng-deep .mat-progress-bar .mat-progress-bar-element{transition:none}"]
                    }] }
        ];
        /** @nocollapse */
        MatFileUploadComponent.ctorParameters = function () { return [
            { type: http.HttpClient },
            { type: MatFileUploadQueueService }
        ]; };
        MatFileUploadComponent.propDecorators = {
            httpUrl: [{ type: core.Input }],
            httpRequestHeaders: [{ type: core.Input }],
            httpRequestParams: [{ type: core.Input }],
            fileAlias: [{ type: core.Input }],
            file: [{ type: core.Input }],
            id: [{ type: core.Input }],
            fileUploadAriaLabel: [{ type: core.Input }],
            cancelAriaLabel: [{ type: core.Input }],
            removeEvent: [{ type: core.Output }],
            onUpload: [{ type: core.Output }]
        };
        return MatFileUploadComponent;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        MatFileUploadComponent.prototype.uploadProgressSubject;
        /** @type {?} */
        MatFileUploadComponent.prototype.uploadProgress$;
        /**
         * @type {?}
         * @private
         */
        MatFileUploadComponent.prototype.uploadInProgressSubject;
        /** @type {?} */
        MatFileUploadComponent.prototype.uploadInProgress$;
        /** @type {?} */
        MatFileUploadComponent.prototype.subs;
        /** @type {?} */
        MatFileUploadComponent.prototype.httpUrl;
        /** @type {?} */
        MatFileUploadComponent.prototype.httpRequestHeaders;
        /** @type {?} */
        MatFileUploadComponent.prototype.httpRequestParams;
        /** @type {?} */
        MatFileUploadComponent.prototype.fileAlias;
        /**
         * @type {?}
         * @private
         */
        MatFileUploadComponent.prototype._file;
        /**
         * @type {?}
         * @private
         */
        MatFileUploadComponent.prototype._id;
        /** @type {?} */
        MatFileUploadComponent.prototype.fileUploadAriaLabel;
        /** @type {?} */
        MatFileUploadComponent.prototype.cancelAriaLabel;
        /**
         * Output
         * @type {?}
         */
        MatFileUploadComponent.prototype.removeEvent;
        /** @type {?} */
        MatFileUploadComponent.prototype.onUpload;
        /**
         * @type {?}
         * @private
         */
        MatFileUploadComponent.prototype.fileUploadSubscription;
        /**
         * @type {?}
         * @private
         */
        MatFileUploadComponent.prototype.HttpClient;
        /**
         * @type {?}
         * @private
         */
        MatFileUploadComponent.prototype.matFileUploadQueueService;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/mat-file-upload-queue/mat-file-upload-queue.component.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MatFileUploadQueueComponent = /** @class */ (function () {
        function MatFileUploadQueueComponent(matFileUploadQueueService, changeDetectorRef) {
            this.matFileUploadQueueService = matFileUploadQueueService;
            this.changeDetectorRef = changeDetectorRef;
            this.files = [];
            this.httpRequestHeaders = new http.HttpHeaders();
            this.httpRequestParams = new http.HttpParams();
            this.fileAlias = "file";
            this.uploadAllColor = "primary";
            this.uploadAllLabel = "Upload All";
            this.removeAllColor = "primary";
            this.removeAllLabel = "Remove All";
        }
        Object.defineProperty(MatFileUploadQueueComponent.prototype, "fileUploadRemoveEvents", {
            /** Combined stream of all of the file upload remove change events. */
            get: /**
             * Combined stream of all of the file upload remove change events.
             * @return {?}
             */
            function () {
                return rxjs.merge.apply(void 0, __spread(this.fileUploads.map((/**
                 * @param {?} fileUpload
                 * @return {?}
                 */
                function (fileUpload) { return fileUpload.removeEvent; }))));
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @param {?} changes
         * @return {?}
         */
        MatFileUploadQueueComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
        function (changes) {
            this.matFileUploadQueueService.initialize({
                httpUrl: changes["httpUrl"] ? changes["httpUrl"].currentValue : undefined,
                httpRequestHeaders: changes["httpRequestHeaders"]
                    ? changes["httpRequestHeaders"].currentValue
                    : undefined,
                httpRequestParams: changes["httpRequestParams"]
                    ? changes["httpRequestParams"].currentValue
                    : undefined,
                fileAlias: changes["fileAlias"]
                    ? changes["fileAlias"].currentValue
                    : undefined,
            });
        };
        /**
         * @return {?}
         */
        MatFileUploadQueueComponent.prototype.ngAfterViewInit = /**
         * @return {?}
         */
        function () {
            var _this = this;
            // When the list changes, re-subscribe
            this._changeSubscription = this.fileUploads.changes
                .pipe(operators.startWith(null))
                .subscribe((/**
             * @return {?}
             */
            function () {
                if (_this._fileRemoveSubscription) {
                    _this._fileRemoveSubscription.unsubscribe();
                }
                _this._listenTofileRemoved();
            }));
        };
        /**
         * @private
         * @return {?}
         */
        MatFileUploadQueueComponent.prototype._listenTofileRemoved = /**
         * @private
         * @return {?}
         */
        function () {
            var _this = this;
            this._fileRemoveSubscription = this.fileUploadRemoveEvents.subscribe((/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                _this.files.splice(event.id, 1);
                _this.changeDetectorRef.markForCheck();
            }));
        };
        /**
         * @param {?} file
         * @return {?}
         */
        MatFileUploadQueueComponent.prototype.add = /**
         * @param {?} file
         * @return {?}
         */
        function (file) {
            this.files.push(file);
            this.changeDetectorRef.markForCheck();
        };
        /**
         * @return {?}
         */
        MatFileUploadQueueComponent.prototype.uploadAll = /**
         * @return {?}
         */
        function () {
            this.fileUploads.forEach((/**
             * @param {?} fileUpload
             * @return {?}
             */
            function (fileUpload) {
                fileUpload.upload();
            }));
        };
        /**
         * @return {?}
         */
        MatFileUploadQueueComponent.prototype.removeAll = /**
         * @return {?}
         */
        function () {
            this.files.splice(0, this.files.length);
            this.changeDetectorRef.markForCheck();
        };
        /**
         * @return {?}
         */
        MatFileUploadQueueComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
        function () {
            if (this._changeSubscription)
                this._changeSubscription.unsubscribe();
            if (this._fileRemoveSubscription)
                this._fileRemoveSubscription.unsubscribe();
            if (this.files) {
                this.removeAll();
            }
        };
        MatFileUploadQueueComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "mat-file-upload-queue",
                        template: "<ng-content></ng-content>\n<br />\n<button\n  mat-raised-button\n  [color]=\"uploadAllColor\"\n  *ngIf=\"files.length > 0\"\n  (click)=\"uploadAll()\"\n>\n  {{ uploadAllLabel }}\n</button>\n<button\n  mat-raised-button\n  [color]=\"removeAllColor\"\n  *ngIf=\"files.length > 0\"\n  (click)=\"removeAll()\"\n>\n  {{ removeAllLabel }}\n</button>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        providers: [MatFileUploadQueueService],
                        styles: [""]
                    }] }
        ];
        /** @nocollapse */
        MatFileUploadQueueComponent.ctorParameters = function () { return [
            { type: MatFileUploadQueueService },
            { type: core.ChangeDetectorRef }
        ]; };
        MatFileUploadQueueComponent.propDecorators = {
            fileUploads: [{ type: core.ContentChildren, args: [core.forwardRef((/**
                         * @return {?}
                         */
                        function () { return MatFileUploadComponent; })),] }],
            httpUrl: [{ type: core.Input }],
            httpRequestHeaders: [{ type: core.Input }],
            httpRequestParams: [{ type: core.Input }],
            fileAlias: [{ type: core.Input }],
            uploadAllColor: [{ type: core.Input }],
            uploadAllLabel: [{ type: core.Input }],
            removeAllColor: [{ type: core.Input }],
            removeAllLabel: [{ type: core.Input }]
        };
        return MatFileUploadQueueComponent;
    }());
    if (false) {
        /** @type {?} */
        MatFileUploadQueueComponent.prototype.fileUploads;
        /**
         * Subscription to remove changes in files.
         * @type {?}
         * @private
         */
        MatFileUploadQueueComponent.prototype._fileRemoveSubscription;
        /**
         * Subscription to changes in the files.
         * @type {?}
         * @private
         */
        MatFileUploadQueueComponent.prototype._changeSubscription;
        /** @type {?} */
        MatFileUploadQueueComponent.prototype.files;
        /** @type {?} */
        MatFileUploadQueueComponent.prototype.httpUrl;
        /** @type {?} */
        MatFileUploadQueueComponent.prototype.httpRequestHeaders;
        /** @type {?} */
        MatFileUploadQueueComponent.prototype.httpRequestParams;
        /** @type {?} */
        MatFileUploadQueueComponent.prototype.fileAlias;
        /** @type {?} */
        MatFileUploadQueueComponent.prototype.uploadAllColor;
        /** @type {?} */
        MatFileUploadQueueComponent.prototype.uploadAllLabel;
        /** @type {?} */
        MatFileUploadQueueComponent.prototype.removeAllColor;
        /** @type {?} */
        MatFileUploadQueueComponent.prototype.removeAllLabel;
        /**
         * @type {?}
         * @private
         */
        MatFileUploadQueueComponent.prototype.matFileUploadQueueService;
        /**
         * @type {?}
         * @private
         */
        MatFileUploadQueueComponent.prototype.changeDetectorRef;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/file-upload-input-for/file-upload-input-for.directive.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * A material design file upload queue component.
     */
    var FileUploadInputForDirective = /** @class */ (function () {
        function FileUploadInputForDirective(element) {
            this.element = element;
            this._queue = null;
            this.onFileSelected = new core.EventEmitter();
            this._element = this.element.nativeElement;
        }
        Object.defineProperty(FileUploadInputForDirective.prototype, "fileUploadQueue", {
            set: /**
             * @param {?} value
             * @return {?}
             */
            function (value) {
                if (value) {
                    this._queue = value;
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        FileUploadInputForDirective.prototype.onChange = /**
         * @return {?}
         */
        function () {
            /** @type {?} */
            var files = this.element.nativeElement.files;
            this.onFileSelected.emit(files);
            for (var i = 0; i < files.length; i++) {
                this._queue.add(files[i]);
            }
            this.element.nativeElement.value = "";
        };
        /**
         * @param {?} event
         * @return {?}
         */
        FileUploadInputForDirective.prototype.onDrop = /**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            /** @type {?} */
            var files = event.dataTransfer.files;
            this.onFileSelected.emit(files);
            for (var i = 0; i < files.length; i++) {
                this._queue.add(files[i]);
            }
            event.preventDefault();
            event.stopPropagation();
            this.element.nativeElement.value = "";
        };
        /**
         * @param {?} event
         * @return {?}
         */
        FileUploadInputForDirective.prototype.onDropOver = /**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            event.preventDefault();
        };
        FileUploadInputForDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: "input[fileUploadInputFor], div[fileUploadInputFor]",
                    },] }
        ];
        /** @nocollapse */
        FileUploadInputForDirective.ctorParameters = function () { return [
            { type: core.ElementRef }
        ]; };
        FileUploadInputForDirective.propDecorators = {
            onFileSelected: [{ type: core.Output }],
            fileUploadQueue: [{ type: core.Input, args: ["fileUploadInputFor",] }],
            onChange: [{ type: core.HostListener, args: ["change",] }],
            onDrop: [{ type: core.HostListener, args: ["drop", ["$event"],] }],
            onDropOver: [{ type: core.HostListener, args: ["dragover", ["$event"],] }]
        };
        return FileUploadInputForDirective;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        FileUploadInputForDirective.prototype._queue;
        /**
         * @type {?}
         * @private
         */
        FileUploadInputForDirective.prototype._element;
        /** @type {?} */
        FileUploadInputForDirective.prototype.onFileSelected;
        /**
         * @type {?}
         * @private
         */
        FileUploadInputForDirective.prototype.element;
    }

    /**
     * @fileoverview added by tsickle
     * Generated from: lib/mat-file-upload.module.ts
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var MatFileUploadModule = /** @class */ (function () {
        function MatFileUploadModule() {
        }
        MatFileUploadModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [
                            BytesPipe,
                            MatFileUploadQueueComponent,
                            MatFileUploadComponent,
                            FileUploadInputForDirective,
                        ],
                        imports: [
                            progressBar.MatProgressBarModule,
                            card.MatCardModule,
                            button.MatButtonModule,
                            icon.MatIconModule,
                            http.HttpClientModule,
                            common.CommonModule,
                        ],
                        exports: [
                            BytesPipe,
                            MatFileUploadQueueComponent,
                            MatFileUploadComponent,
                            FileUploadInputForDirective,
                        ],
                    },] }
        ];
        return MatFileUploadModule;
    }());

    exports.FileUploadInputForDirective = FileUploadInputForDirective;
    exports.MatFileUploadComponent = MatFileUploadComponent;
    exports.MatFileUploadModule = MatFileUploadModule;
    exports.MatFileUploadQueueComponent = MatFileUploadQueueComponent;
    exports.ɵa = BytesPipe;
    exports.ɵb = MatFileUploadQueueService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
