// igniteui-angular/esm2015/lib/tabs/tab-item.directive.js
import { ContentChild, Directive, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { Direction } from '../carousel/carousel-base';
import { IgxTabHeaderBase, IgxTabContentBase, IgxTabsBase } from './tabs.base';
export class IgxTabItemDirective {
    /** @hidden */
    constructor(tabs) {
        this.tabs = tabs;
        /**
         * Output to enable support for two-way binding on [(selected)]
         */
        this.selectedChange = new EventEmitter();
        /**
         * An @Input property that allows you to enable/disable the item.
         */
        this.disabled = false;
        /** @hidden */
        this.direction = Direction.NONE;
        this._selected = false;
    }
    /**
     * An @Input property which determines whether an item is selected.
     */
    get selected() {
        return this._selected;
    }
    ;
    set selected(value) {
        if (this._selected !== value) {
            this._selected = value;
            this.tabs.selectTab(this, this._selected);
            this.selectedChange.emit(this._selected);
        }
    }
}
IgxTabItemDirective.decorators = [
    { type: Directive }
];
IgxTabItemDirective.ctorParameters = () => [
    { type: IgxTabsBase }
];
IgxTabItemDirective.propDecorators = {
    headerComponent: [{ type: ContentChild, args: [IgxTabHeaderBase,] }],
    panelComponent: [{ type: ContentChild, args: [IgxTabContentBase,] }],
    headerTemplate: [{ type: ViewChild, args: ['headerTemplate', { static: true },] }],
    panelTemplate: [{ type: ViewChild, args: ['panelTemplate', { static: true },] }],
    selectedChange: [{ type: Output }],
    disabled: [{ type: Input }],
    selected: [{ type: Input }]
};
